[ML] IndexOutOfBoundsException in frequent items agg

This exception was observed in a Cloud cluster:

```
Failed to execute phase [query], all shards failed; shardFailures {[A6izEymCQDqJ1Ejc2tj1Nw][so][0]: org.elasticsearch.transport.RemoteTransportException: [es-es-search-6fb557dcd5-t2gl4][10.2.145.120:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.search.query.QueryPhaseExecutionException: Query Failed [Failed to execute main query]
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:229)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.executeQuery(QueryPhase.java:135)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:63)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:518)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:671)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:543)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:51)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:48)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at java.base/java.util.Collections$UnmodifiableList.get(Collections.java:1501)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.mr.ItemSetMapReduceAggregator.getLeafReaderForOrdinals(ItemSetMapReduceAggregator.java:225)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.mr.ItemSetMapReduceAggregator.<init>(ItemSetMapReduceAggregator.java:79)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.FrequentItemSetsAggregatorFactory$1.<init>(FrequentItemSetsAggregatorFactory.java:143)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.FrequentItemSetsAggregatorFactory.createInternal(FrequentItemSetsAggregatorFactory.java:129)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactory.create(AggregatorFactory.java:80)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactories.createSubAggregators(AggregatorFactories.java:233)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactories.createTopLevelAggregators(AggregatorFactories.java:243)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregationPhase.newAggregatorCollector(AggregationPhase.java:51)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregationPhase.lambda$preProcess$1(AggregationPhase.java:37)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorCollectorManager.newCollector(AggregatorCollectorManager.java:39)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorCollectorManager.newCollector(AggregatorCollectorManager.java:21)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhaseCollectorManager.newCollector(QueryPhaseCollectorManager.java:139)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:311)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:205)
	... 15 more
}
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:709)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:401)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:741)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:494)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:332)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:553)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1703)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1420)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1556)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1531)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:51)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:37)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:124)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.elasticsearch.ElasticsearchException$1: Index 0 out of bounds for length 0
	at org.elasticsearch.server@8.12.0/org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:672)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:399)
	... 22 more
Caused by: java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at java.base/java.util.Collections$UnmodifiableList.get(Collections.java:1501)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.mr.ItemSetMapReduceAggregator.getLeafReaderForOrdinals(ItemSetMapReduceAggregator.java:225)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.mr.ItemSetMapReduceAggregator.<init>(ItemSetMapReduceAggregator.java:79)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.FrequentItemSetsAggregatorFactory$1.<init>(FrequentItemSetsAggregatorFactory.java:143)
	at org.elasticsearch.ml@8.12.0/org.elasticsearch.xpack.ml.aggs.frequentitemsets.FrequentItemSetsAggregatorFactory.createInternal(FrequentItemSetsAggregatorFactory.java:129)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactory.create(AggregatorFactory.java:80)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactories.createSubAggregators(AggregatorFactories.java:233)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorFactories.createTopLevelAggregators(AggregatorFactories.java:243)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregationPhase.newAggregatorCollector(AggregationPhase.java:51)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregationPhase.lambda$preProcess$1(AggregationPhase.java:37)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorCollectorManager.newCollector(AggregatorCollectorManager.java:39)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.aggregations.AggregatorCollectorManager.newCollector(AggregatorCollectorManager.java:21)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhaseCollectorManager.newCollector(QueryPhaseCollectorManager.java:139)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:311)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.addCollectorsAndSearch(QueryPhase.java:205)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.executeQuery(QueryPhase.java:135)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:63)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:518)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:671)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:543)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:51)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:48)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
	at org.elasticsearch.server@8.12.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 6 more
```

The suspicion is that there is no data and the aggregation is assuming that there is at least 1 segment.

--- a/docs/changelog/103116.yaml
+++ b/docs/changelog/103116.yaml
@@ -1,0 +1,6 @@
+pr: 103116
+summary: Fix `frequent_item_sets` aggregation on empty index
+area: Machine Learning
+type: bug
+issues:
+ - 103067

--- a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/mr/ItemSetMapReduceAggregator.java
+++ b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/mr/ItemSetMapReduceAggregator.java
@@ -42,6 +42,7 @@ import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.function.BiConsumer;
 
 public abstract class ItemSetMapReduceAggregator<
@@ -76,7 +77,7 @@ public abstract class ItemSetMapReduceAggregator<
         List<ItemSetMapReduceValueSource> valueSources = new ArrayList<>();
         List<Field> fields = new ArrayList<>();
         IndexSearcher contextSearcher = context.searcher();
-        LeafReaderContext ctx = getLeafReaderForOrdinals(context);
+        Optional<LeafReaderContext> ctx = getLeafReaderForOrdinals(context);
 
         int id = 0;
         this.weightDocumentFilter = documentFilter != null
@@ -85,15 +86,17 @@ public abstract class ItemSetMapReduceAggregator<
 
         boolean rewriteBasedOnOrdinals = false;
 
-        for (var c : configsAndValueFilters) {
-            ItemSetMapReduceValueSource e = context.getValuesSourceRegistry()
-                .getAggregator(registryKey, c.v1())
-                .build(c.v1(), id++, c.v2(), ordinalOptimization, ctx);
-            if (e.getField().getName() != null) {
-                fields.add(e.getField());
-                valueSources.add(e);
+        if (ctx.isPresent()) {
+            for (var c : configsAndValueFilters) {
+                ItemSetMapReduceValueSource e = context.getValuesSourceRegistry()
+                    .getAggregator(registryKey, c.v1())
+                    .build(c.v1(), id++, c.v2(), ordinalOptimization, ctx.get());
+                if (e.getField().getName() != null) {
+                    fields.add(e.getField());
+                    valueSources.add(e);
+                }
+                rewriteBasedOnOrdinals |= e.usesOrdinals();
             }
-            rewriteBasedOnOrdinals |= e.usesOrdinals();
         }
 
         this.rewriteBasedOnOrdinals = rewriteBasedOnOrdinals;
@@ -220,8 +223,8 @@ public abstract class ItemSetMapReduceAggregator<
         return new InternalItemSetMapReduceAggregation<>(name, metadata(), mapReducer, context, null, fields, profiling);
     }
 
-    private static LeafReaderContext getLeafReaderForOrdinals(AggregationContext context) {
+    private static Optional<LeafReaderContext> getLeafReaderForOrdinals(AggregationContext context) {
         IndexReader reader = context.searcher().getIndexReader();
-        return reader.leaves().get(0);
+        return reader.leaves().stream().findFirst();
     }
 }

--- a/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
+++ b/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
@@ -550,7 +550,7 @@ setup:
       headers:
         Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
       indices.create:
-        index: empty_store
+        index: unavailable-data
         body:
           mappings:
             properties:
@@ -567,7 +567,7 @@ setup:
 
   - do:
       search:
-        index: empty_store
+        index: unavailable-data
         body: >
           {
             "size": 0,

--- a/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
+++ b/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
@@ -544,9 +544,7 @@ setup:
 ---
 "Test frequent items on empty index":
   - skip:
-      features:
-        - "headers"
-        - "allowed_warnings"
+      features: headers
 
   - do:
       headers:

--- a/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
+++ b/x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/frequent_item_sets_agg.yml
@@ -540,3 +540,50 @@ setup:
   - match: { aggregations.fi.buckets.1.doc_count: 4 }
   - match: { aggregations.fi.buckets.1.support: 0.4 }
   - match: { aggregations.fi.buckets.1.key.error_message: ["engine overheated"] }
+
+---
+"Test frequent items on empty index":
+  - skip:
+      features:
+        - "headers"
+        - "allowed_warnings"
+
+  - do:
+      headers:
+        Authorization: "Basic eF9wYWNrX3Jlc3RfdXNlcjp4LXBhY2stdGVzdC1wYXNzd29yZA==" # run as x_pack_rest_user, i.e. the test setup superuser
+      indices.create:
+        index: empty_store
+        body:
+          mappings:
+            properties:
+              features:
+                type: keyword
+              error_message:
+                type: keyword
+              timestamp:
+                type: date
+              geo_point:
+                type: geo_point
+              histogram:
+                type: histogram
+
+  - do:
+      search:
+        index: empty_store
+        body: >
+          {
+            "size": 0,
+            "aggs": {
+              "fi": {
+                "frequent_item_sets": {
+                  "minimum_set_size": 3,
+                  "minimum_support": 0.3,
+                  "fields": [
+                    {"field": "features"},
+                    {"field": "error_message"}
+                  ]
+                }
+              }
+            }
+          }
+  - length: { aggregations.fi.buckets: 0 }

